This rule is a stub.
## Noncompliant Code Example
This noncompliant code example shows an example where ...
``` java
```
## Compliant Solution
In this compliant solution, ...
``` java
```
## Risk Assessment
Using insecure or weak cryptographic algorithms is not a good idea.

| Rule | Severity | Likelihood | Remediation Cost | Priority | Level |
| ----|----|----|----|----|----|
| MSC25-C | Medium | Probable | High | P4 | L3 |

## Automated Detection

| Tool | Version | Checker | Description |
| ----|----|----|----|
| CodeSonar | 8.3p0 | BADMACRO.WEAK_CRYPTOBADFUNC.WEAKCRYPTO | Use of Weak Cryptographic AlgorithmWeak Cryptography |

## Related Guidelines

|  |  |
| ----|----|
| SEI CERT Oracle Coding Standard for Java | MSC61-J. Do not use insecure or weak cryptographic algorithms |

## Bibliography
